github/prci.py

```diff
@@ -247,49 +217,80 @@ def main():
+    whitelist = config["whitelist"]
+    no_task_backoff_time = config["no_task_backoff_time"]
+
+    logging.config.dictConfig(config["logging"])
+
+    exit_handler = ExitHandler()
+    signal.signal(signal.SIGINT, exit_handler.finish)
+    signal.signal(signal.SIGTERM, exit_handler.abort)
+
+    gh = github3.login(token=credentials["token"])
+    session = util.create_session(util.make_headers(credentials["token"]))
+    do_request = partial(util.perform_request, session=session)
+
+    world = World(
+        graphql_request=do_request,
+        github_api=gh,
+        session=session,
+        repo_owner=repo["owner"],
+        repo_name=repo["name"],
+        runner_id=runner_id,
+        tasks_path=tasks_path,
+        whitelist=whitelist
+    )
 
-    handler = ExitHandler()
-    signal.signal(signal.SIGINT, handler.finish)
-    signal.signal(signal.SIGTERM, handler.abort)
+    while not exit_handler.done:
+        world.check_graphql_limit()
 
-    while not handler.done:
         try:
-            task_queue.create_tasks_for_pulls()
-
-            try:
-                task = next(task_queue)
-            except StopIteration:
-                time.sleep(no_task_backoff_time)
-                continue
-
-            handler.register_task(task)
-            task.execute()
-        except Exception:
-            sentry_report_exception({'module': 'github'})
-            time.sleep(ERROR_BACKOFF_TIME)
-        finally:
-            handler.unregister_task()
-
-            try:
-                check_reboot(repo, creds)
-            except:
-                sentry_report_exception()
-
-
-if __name__ == '__main__':
+            response = do_request(
+                query=queries.make_pull_requests_query(
+                    world.repo_owner, world.repo_name
+                )
+            )
+        except EnvironmentError as e:
+            logger.error(e)
+            sys.exit(1)
+
+        data = util.get_data(response)
+        repo = util.get_repository(data)
+        repo_url = util.get_repository_url(repo)
+        pull_requests_data = util.get_pull_requests(repo)
+
+        pull_requests = sorted(
+            (
+                PullRequest.from_dict(pr_data)
+                for pr_data in pull_requests_data
+            ),
+            key=lambda pr: not pr.prioritized
+        )
+        for pull_request in pull_requests:
+            for task in process_pull_request(world, pull_request, repo_url):
+                exit_handler.register_task(task)
+                world.available_resources.take(task)
+                logger.info(
+                    "Available resources: %s", world.available_resources
+                )
+                try:
+                    task.execute(world, pull_request.commit.statuses)
+                except (EnvironmentError, RuntimeError) as e:
+                    logger.error(e)
+                    sentry_report_exception({"module": "github"})
+                    sleep(ERROR_BACKOFF_TIME)
```

Comment:
I think we can get rid of this. IIUC the purpose of this in the first place was to avoid hitting the API rate limit.

Reply:
But we will reach the rate limit very quickly if we will not wait for some time anyway.